src/modules/flow/oauth/oauth.c

```diff
@@ -0,0 +1,596 @@
+        HTTP_SERVER_PORT, namespace);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    r = sol_http_escape_string(&url, temp);
+    free(temp);
+    SOL_INT_CHECK(r, < 0, NULL);
+
+    return url;
+}
+
+static void
+digest_ready_cb(void *data, struct sol_message_digest *handle, struct sol_blob *output)
+{
+    int r;
+    struct sol_http_param params;
+    struct sol_http_client_connection *connection;
+    struct v1_request_data *req_data = data;
+    struct v1_data *mdata = req_data->mdata;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+
+    r = sol_buffer_append_as_base64(&buffer, sol_str_slice_from_blob(output), SOL_BASE64_MAP);
+    SOL_INT_CHECK_GOTO(r, < 0, err);
+
+    sol_http_param_init(&params);
+    if (!sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_HEADER("oauth_callback", mdata->callback_url)) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_consumer_key", mdata->consumer_key)) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_nonce", req_data->nonce)) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_signature_method", "HMAC-SHA1")) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_timestamp", req_data->timestamp)) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_version", "1.0")) ||
+        !sol_http_param_add(&params,
+        SOL_HTTP_REQUEST_PARAM_POST_FIELD("oauth_signature", buffer.data))) {
+        SOL_WRN("Failed to set query params");
+        sol_http_param_free(&params);
+        sol_buffer_fini(&buffer);
+        goto err;
+    }
+
+    connection = sol_http_client_request(SOL_HTTP_METHOD_POST, mdata->request_token_url,
+        &params, v1_request_finished, req_data);
+    sol_buffer_fini(&buffer);
+    sol_http_param_free(&params);
+    SOL_NULL_CHECK_GOTO(connection, err);
+
+    r = sol_ptr_vector_append(&mdata->pending_conns, connection);
+    SOL_INT_CHECK_GOTO(r, < 0, err_append);
+    return;
+
+err_append:
+    sol_http_client_connection_cancel(connection);
+err:
+    free(req_data->timestamp);
+    free(req_data->nonce);
+    r = sol_http_server_send_response(req_data->request, &((struct sol_http_response) {
+            .api_version = SOL_HTTP_RESPONSE_API_VERSION,
+            .content = SOL_BUFFER_INIT_EMPTY,
+            .param = SOL_HTTP_REQUEST_PARAM_INIT,
+            .response_code = SOL_HTTP_STATUS_INTERNAL_SERVER_ERROR
+        }));
+    if (r < 0) {
+        SOL_WRN("Could not send fail response for %s",
+            sol_http_request_get_url(req_data->request));
+    }
+    free(req_data);
+}
+
+static char *
+generate_nonce(void)
+{
+    int r;
+    ssize_t size;
+    char *nonce = NULL;
+    struct sol_random *engine;
+    struct sol_str_slice slice;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+
+    engine = sol_random_new(SOL_RANDOM_DEFAULT, 0);
```

Comment:
For nonces, I'd use `SOL_RANDOM_URANDOM`. This falls back to `MT19937` (the default) if `/dev/urandom` can't be opened for any reason (which, if available, will initialize the seed using `getrandom()` or `/dev/urandom`, but the seed has only 4 bytes).